### ISSUE TITLE
Reading unsigned `grad_fn._saved_dim` for dim=-1 leads to an incorrect comparison with `MAX_VALID_DIM`  and an infinite loop for `cat` operations in RoPE blocks in .get_pruning_group()

### ISSUE THREAD

**Issue by Martakor1** [VainF/Torch-Pruning#532]
## Description

`torch_pruning.dependency.index_mapping.update_concat_index_mapping()` reads `cat_node.grad_fn._saved_dim` and checks it against `constants.MAX_VALID_DIM`:

```python
if (
    hasattr(cat_node.grad_fn, "_saved_dim")
    and cat_node.grad_fn._saved_dim != 1
    and cat_node.grad_fn._saved_dim < constants.MAX_VALID_DIM
):
    return
```

In some PyTorch versions, a negative dimension such as `dim=-1` is exposed through the private autograd field `_saved_dim` as an unsigned 64-bit value:

```text
18446744073709551615
```

This value is the unsigned representation of `-1`:

```text
18446744073709551615 == 2**64 - 1
```

Because `constants.MAX_VALID_DIM` is currently `100`, the comparison:

```python
cat_node.grad_fn._saved_dim < constants.MAX_VALID_DIM
```

returns `False`.

As a result, `update_concat_index_mapping()` does not return and incorrectly creates a `_ConcatIndexMapping` for a concatenation whose actual dimension is `-1`.

In dependency graphs containing cycles, such as attention/RoPE graphs, the incorrect mapping can repeatedly shift pruning indices. Every traversal then appears to be a new pruning state, preventing `get_pruning_group()` from terminating.

## Minimal example demonstrating `_saved_dim`

```python
import torch

x = torch.randn(2, 3, 4, 5, requires_grad=True)

y = torch.cat(
    (x[..., :2], x[..., 2:]),
    dim=-1,
)

print(y.grad_fn)
print(y.grad_fn._saved_dim)
print(type(y.grad_fn._saved_dim))
```

Output with PyTorch `2.10.0+cu128`:

```text
<CatBackward0 object at ...>
18446744073709551615
<class 'int'>
```

Although the original operation uses `dim=-1`, `_saved_dim` is returned as the Python integer `18446744073709551615`.

The same behavior can also be observed with operations such as `torch.split(..., dim=-1)`.

## Observed infinite loop

The problem appeared while building a pruning group for a projection inside an attention block:

```python
group = DG.get_pruning_group(
    q_proj,
    tp.prune_linear_out_channels,
    idxs=[2, 6, 9],
)
```

The dependency traversal reached a RoPE operation similar to:

```python
def rotate_half(x):
    x1 = x[..., : x.shape[-1] // 2]
    x2 = x[..., x.shape[-1] // 2 :]
    return torch.cat((-x2, x1), dim=-1)
```

The corresponding concat node received an incorrect `_ConcatIndexMapping`. On every traversal of the cycle, its offsets were applied again, producing index sequences similar to:

```text
[2, 6, 9]
[1026, 1030, 1033]
[2050, 2054, 2057]
[3074, 3078, 3081]
[4098, 4102, 4105]
...
```

`DependencyGraph.get_pruning_group()` attempts to terminate revisits with:

```python
if (
    new_dep.target in visited_node
    and group.has_pruning_op(new_dep, new_indices)
):
    continue
```

However, because `new_indices` change on every traversal, `group.has_pruning_op()` continues returning `False`. The dependency is added to the processing stack again, causing effectively infinite traversal and unbounded growth of the pruning group.

In debugging, the traversal produced more than `100,000` unique dependency states even though the dependency graph itself contained only a few thousand nodes.

## Temporary workaround

The issue can currently be avoided by changing:

```python
MAX_VALID_DIM = 100
```

to:

```python
MAX_VALID_DIM = 1 << 64
```

With this value:

```python
(2**64 - 1) < (2**64)
```

evaluates to `True`, so the existing condition correctly returns for this particular case.

However, this workaround relies on the unsigned representation detail and should not be considered a robust fix.

## Suggested fix

`_saved_dim` should be normalized from its unsigned 64-bit representation to a signed integer before it is interpreted:

```python
saved_dim = cat_node.grad_fn._saved_dim

if saved_dim >= (1 << 63):
    saved_dim -= 1 << 64
```

After normalization:

```python
18446744073709551615 → -1
```

The existing comparison can then operate on the actual dimension:

```python
if (
    hasattr(cat_node.grad_fn, "_saved_dim")
    and saved_dim != 1
    and saved_dim < constants.MAX_VALID_DIM
):
    return
```

Alternatively, the dimension could be normalized against the rank of the relevant tensor:

```python
if saved_dim < 0:
    saved_dim += ndim
```

The important part is that `18446744073709551615` should not be treated as an ordinary large positive dimension. It represents `-1` in this case.

Since `_saved_dim` is a private PyTorch autograd attribute, it may also be worth isolating its decoding in a helper and handling unexpected representations defensively.

## Environment

```text
torch: 2.10.0+cu128
torch-pruning: 1.6.0
GPU: NVIDIA A100-PCIE-40GB
```

## Expected behavior

`torch.cat(..., dim=-1)` should be recognized as concatenation along dimension `-1`, and Torch-Pruning should not install an incorrect channel mapping because the dimension was interpreted as `2**64 - 1`.

`DependencyGraph.get_pruning_group()` should terminate normally when such a concat operation occurs inside a cyclic attention/RoPE dependency graph.

## Actual behavior

The unsigned representation of `_saved_dim` bypasses the `MAX_VALID_DIM` check, an incorrect concat index mapping is installed, pruning indices change after every cycle, and `get_pruning_group()` does not terminate.
